Some tests error out with typing_extensions>=4.12

* mashumaro version: 3.13 (bffc024)
* Python version: 3.12, 3.11
* Operating System: Linux (WSL)

### Description

Trying to run the tests with typing_extensions>=4.12 (released 3 weeks ago) results in about 10 errors with message `mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro`. With typing_extensions==4.11 the tests are collected without errors.

The relevant changelog entry is:
> Backport the typing.NoDefault sentinel object from Python 3.13. TypeVars, ParamSpecs and TypeVarTuples without default values now have their __default__ attribute set to this sentinel value.

### What I Did

```
pytest
=========================================== test session starts ============================================
platform linux -- Python 3.11.0rc1, pytest-8.2.0, pluggy-1.5.0
rootdir: /home/danzin/projects/mashumaro
configfile: pyproject.toml
plugins: mock-3.14.0, time-machine-2.14.1, flaky-3.8.1, hypothesis-6.100.2, xdist-3.5.0, cov-5.0.0
collected 318 items / 10 errors

================================================== ERRORS ==================================================
__________________________ ERROR collecting tests/test_codecs/test_basic_codec.py __________________________
tests/test_codecs/test_basic_codec.py:11: in <module>
    from tests.entities import (
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
__________________________________ ERROR collecting tests/test_config.py ___________________________________
tests/test_config.py:12: in <module>
    from .entities import (
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
________________________________ ERROR collecting tests/test_data_types.py _________________________________
tests/test_data_types.py:63: in <module>
    from tests.entities import MyUntypedNamedTupleWithDefaults
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
_________________________________ ERROR collecting tests/test_generics.py __________________________________
tests/test_generics.py:7: in <module>
    from tests.entities import MyGenericDataClass, SerializableTypeGenericList
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
___________________________________ ERROR collecting tests/test_json.py ____________________________________
tests/test_json.py:11: in <module>
    from .entities import MyEnum
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
___________________ ERROR collecting tests/test_jsonschema/test_jsonschema_generation.py ___________________
tests/test_jsonschema/test_jsonschema_generation.py:77: in <module>
    from tests.entities import (
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
__________________________________ ERROR collecting tests/test_literal.py __________________________________
tests/test_literal.py:11: in <module>
    from tests.entities import MyEnum
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
___________________________________ ERROR collecting tests/test_meta.py ____________________________________
tests/test_meta.py:63: in <module>
    from .entities import (
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
_____________________________ ERROR collecting tests/test_metadata_options.py ______________________________
tests/test_metadata_options.py:18: in <module>
    from .entities import (
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
___________________________________ ERROR collecting tests/test_types.py ___________________________________
tests/test_types.py:12: in <module>
    from tests.entities import GenericSerializableList, GenericSerializableWrapper
tests/entities.py:207: in <module>
    class MyGenericDataClass(Generic[T], DataClassDictMixin):
/usr/lib/python3.11/typing.py:1823: in __init_subclass__
    super().__init_subclass__(*args, **kwargs)
mashumaro/mixins/dict.py:25: in __init_subclass__
    compile_mixin_unpacker(cls, **builder_params["unpacker"])
mashumaro/core/meta/mixin.py:49: in compile_mixin_unpacker
    builder.add_unpack_method()
mashumaro/core/meta/code/builder.py:586: in add_unpack_method
    self._add_unpack_method_lines(method_name)
mashumaro/core/meta/code/builder.py:491: in _add_unpack_method_lines
    ).build(
mashumaro/core/meta/code/builder.py:1332: in build
    unpacked_value = UnpackerRegistry.get(
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:756: in unpack_special_typing_primitive
    uv = UnpackerRegistry.get(spec.copy(type=bound))
mashumaro/core/meta/types/common.py:238: in get
    expr = packer(spec)
mashumaro/core/meta/types/unpack.py:830: in unpack_special_typing_primitive
    raise UnserializableDataError(
E   mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not supported by mashumaro
========================================= short test summary info ==========================================
ERROR tests/test_codecs/test_basic_codec.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_config.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_data_types.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_generics.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_json.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_jsonschema/test_jsonschema_generation.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_literal.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_meta.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_metadata_options.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
ERROR tests/test_types.py - mashumaro.exceptions.UnserializableDataError: typing_extensions.NoDefault as a field type is not suppor...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 10 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================ 10 errors in 4.80s ============================================
```

--- a/mashumaro/core/meta/helpers.py
+++ b/mashumaro/core/meta/helpers.py
@@ -272,8 +272,9 @@ def type_name(
             )
             return f"{_typing_name('Union', short)}[{args_str}]"
         else:
-            bound = getattr(typ, "__default__", None)
-            if bound is None:
+            if type_var_has_default(typ):
+                bound = get_type_var_default(typ)
+            else:
                 bound = getattr(typ, "__bound__")
             return type_name(bound, short, resolved_type_params)
     elif is_new_type(typ) and not PY_310_MIN:
@@ -423,7 +424,7 @@ def is_type_var_any(typ: Type) -> bool:
         return False
     elif typ.__bound__ not in (None, Any):
         return False
-    elif getattr(typ, "__default__", None) not in (None, NoneType):
+    elif type_var_has_default(typ):
         return False
     else:
         return True
@@ -806,3 +807,14 @@ def is_type_alias_type(typ: Type) -> bool:
         return isinstance(typ, typing.TypeAliasType)  # type: ignore
     else:
         return False
+
+
+def type_var_has_default(typ: Any) -> bool:
+    try:
+        return typ.has_default()
+    except AttributeError:
+        return getattr(typ, "__default__", None) is not None
+
+
+def get_type_var_default(typ: Any) -> Type:
+    return getattr(typ, "__default__")

--- a/mashumaro/core/meta/types/pack.py
+++ b/mashumaro/core/meta/types/pack.py
@@ -31,6 +31,7 @@ from mashumaro.core.meta.helpers import (
     get_function_return_annotation,
     get_literal_values,
     get_type_origin,
+    get_type_var_default,
     is_final,
     is_generic,
     is_literal,
@@ -52,6 +53,7 @@ from mashumaro.core.meta.helpers import (
     resolve_type_params,
     substitute_type_params,
     type_name,
+    type_var_has_default,
 )
 from mashumaro.core.meta.types.common import (
     Expression,
@@ -471,8 +473,9 @@ def pack_special_typing_primitive(spec: ValueSpec) -> Optional[Expression]:
             if constraints:
                 return pack_union(spec, constraints, "type_var")
             else:
-                bound = getattr(spec.type, "__default__", None)
-                if bound is None:
+                if type_var_has_default(spec.type):
+                    bound = get_type_var_default(spec.type)
+                else:
                     bound = getattr(spec.type, "__bound__")
                 # act as if it was Optional[bound]
                 pv = PackerRegistry.get(spec.copy(type=bound))

--- a/mashumaro/core/meta/types/unpack.py
+++ b/mashumaro/core/meta/types/unpack.py
@@ -37,6 +37,7 @@ from mashumaro.core.meta.helpers import (
     get_class_that_defines_method,
     get_function_arg_annotation,
     get_literal_values,
+    get_type_var_default,
     is_final,
     is_generic,
     is_literal,
@@ -59,6 +60,7 @@ from mashumaro.core.meta.helpers import (
     resolve_type_params,
     substitute_type_params,
     type_name,
+    type_var_has_default,
 )
 from mashumaro.core.meta.types.common import (
     AbstractMethodBuilder,
@@ -749,8 +751,9 @@ def unpack_special_typing_primitive(spec: ValueSpec) -> Optional[Expression]:
             if constraints:
                 return TypeVarUnpackerBuilder(constraints).build(spec)
             else:
-                bound = getattr(spec.type, "__default__", None)
-                if bound is None:
+                if type_var_has_default(spec.type):
+                    bound = get_type_var_default(spec.type)
+                else:
                     bound = getattr(spec.type, "__bound__")
                 # act as if it was Optional[bound]
                 uv = UnpackerRegistry.get(spec.copy(type=bound))

--- a/tests/test_meta.py
+++ b/tests/test_meta.py
@@ -29,6 +29,7 @@ from mashumaro.core.meta.helpers import (
     get_literal_values,
     get_type_annotations,
     get_type_origin,
+    get_type_var_default,
     hash_type_args,
     is_annotated,
     is_dataclass_dict_mixin,
@@ -49,6 +50,7 @@ from mashumaro.core.meta.helpers import (
     resolve_type_params,
     substitute_type_params,
     type_name,
+    type_var_has_default,
 )
 from mashumaro.core.meta.types.common import (
     FieldContext,
@@ -807,3 +809,25 @@ def test_is_hashable_type():
     assert is_hashable_type(int) is True
     assert is_hashable_type(MyFrozenDataClass) is True
     assert is_hashable_type(MyDataClass) is False
+
+
+def test_type_var_has_default():
+    T_WithoutDefault = typing_extensions.TypeVar("T_WithoutDefault")
+    T_WithDefault = typing_extensions.TypeVar("T_WithDefault", default=int)
+    T_WithDefaultNone = typing_extensions.TypeVar(
+        "T_WithDefaultNone", default=None
+    )
+
+    assert not type_var_has_default(T_WithoutDefault)
+    assert type_var_has_default(T_WithDefault)
+    assert type_var_has_default(T_WithDefaultNone)
+
+
+def test_get_type_var_default():
+    T_WithDefault = typing_extensions.TypeVar("T_WithDefault", default=int)
+    T_WithDefaultNone = typing_extensions.TypeVar(
+        "T_WithDefaultNone", default=None
+    )
+
+    assert get_type_var_default(T_WithDefault) is int
+    assert get_type_var_default(T_WithDefaultNone) is None

